File: templates/production-adapters/opencode-http-session/typescript/prompt-assembly.ts
import { readFile, rm } from "node:fs/promises"
import { join, relative, resolve } from "node:path"

import type { InjectionSection, StageContract, WorkflowBundle } from "./contracts"
import { writeJson, writeText } from "./json-store"

const WORKFLOW_LOCAL_PREFIXES = ["rules/", "prompts/", "reference/"]

export async function renderPrompt(input: {
  readonly attempt: number
  readonly bundle: WorkflowBundle
  readonly repoRoot: string
  readonly runDirectory: string
  readonly stage: string
}): Promise<{ readonly promptPath: string; readonly text: string }> {
  const stageContract = input.bundle.manifest.stages[input.stage]
  if (stageContract === undefined) {
    throw new Error(`unknown stage: ${input.stage}`)
  }
  const promptTemplatePath = join(input.bundle.workflowRoot, stageContract.promptTemplate)
  const sections: string[] = [renderRunnerContext(input.stage, stageContract, input.repoRoot, input.runDirectory)]
  const selectedArtifacts: { label: string; path: string }[] = []
  for (const label of stageContract.injection.renderOrder) {
    const section = findSection(stageContract, label, input.stage)
    const rendered = await readSectionBody({
      bundle: input.bundle,
      promptTemplatePath,
      repoRoot: input.repoRoot,
      runDirectory: input.runDirectory,
      section,
    })
    if (rendered === undefined) {
      continue
    }
    sections.push(`# ${section.label}\n\n${rendered.body}`)
    if (rendered.path !== undefined) {
      selectedArtifacts.push({ label: section.label, path: displayPath(rendered.path, input.repoRoot) })
    }
  }
  if (stageContract.continuationMessage !== undefined) {
    sections.push(`# Continuation Message\n\n${stageContract.continuationMessage}`)
  }
  const promptPath = join(input.runDirectory, "artifacts", "prompts", `${input.stage}-${input.attempt}.md`)
  const text = `${sections.filter((section) => section.trim().length > 0).join("\n\n")}\n`
  await writeText(promptPath, text)
  await writeJson(join(input.runDirectory, stageContract.persistence.selectedArtifactsRecord), selectedArtifacts)
  return { promptPath, text }
}

export async function clearStageOutputs(input: {
  readonly bundle: WorkflowBundle
  readonly repoRoot: string
  readonly runDirectory: string
  readonly stage: string
}): Promise<void> {
  const stageContract = input.bundle.manifest.stages[input.stage]
  if (stageContract === undefined) {
    throw new Error(`unknown stage: ${input.stage}`)
  }
  for (const spec of stageContract.clearOnEnter) {
    await rm(resolveContractPath(spec, input.bundle, input.repoRoot, input.runDirectory), { force: true, recursive: true })
  }
}

export async function ensureStageOutputs(input: {
  readonly bundle: WorkflowBundle
  readonly repoRoot: string
  readonly runDirectory: string
  readonly stage: string
}): Promise<void> {
  const missing: string[] = []
  for (const path of requiredOutputPaths(input)) {
    if (!(await Bun.file(path).exists())) {
      missing.push(path)
    }
  }
  if (missing.length > 0) {
    throw new Error(`stage ${input.stage} is missing required outputs: ${missing.join(", ")}`)
  }
}

export function requiredOutputPaths(input: {
  readonly bundle: WorkflowBundle
  readonly repoRoot: string
  readonly runDirectory: string
  readonly stage: string
}): string[] {
  const stageContract = input.bundle.manifest.stages[input.stage]
  if (stageContract === undefined) {
    throw new Error(`unknown stage: ${input.stage}`)
  }
  return stageContract.requiredOutputs.map((spec) => resolveContractPath(spec, input.bundle, input.repoRoot, input.runDirectory))
}

function resolveContractPath(spec: string, bundle: WorkflowBundle, repoRoot: string, runDirectory: string): string {
  if (WORKFLOW_LOCAL_PREFIXES.some((prefix) => spec.startsWith(prefix))) {
    return join(bundle.workflowRoot, spec)
  }
  if (spec.startsWith("/")) {
    return spec
  }
  return join(runDirectory, spec)
}

function renderRunnerContext(stage: string, stageContract: StageContract, repoRoot: string, runDirectory: string): string {
  return [
    "## Runner Context",
    "",
    `- Stage key: \`${stage}\``,
    `- Repository root: \`${repoRoot}\``,
    `- Run directory: \`${runDirectory}\``,
    "- Required outputs:",
    ...stageContract.requiredOutputs.map((output) => `- ${output}`),
  ].join("\n")
}

async function readSectionBody(input: {
  readonly bundle: WorkflowBundle
  readonly promptTemplatePath: string
  readonly repoRoot: string
  readonly runDirectory: string
  readonly section: InjectionSection
}): Promise<{ readonly body: string; readonly path?: string } | undefined> {
  if (input.section.promptTemplate === true) {
    return { body: await readFile(input.promptTemplatePath, "utf8") }
  }
  if (input.section.artifact === undefined) {
    throw new Error(`Section '${input.section.label}' must define artifact or promptTemplate`)
  }
  const artifactPath = resolveContractPath(input.section.artifact, input.bundle, input.repoRoot, input.runDirectory)
  if (await Bun.file(artifactPath).exists()) {
    return { body: await formatArtifactBody(artifactPath), path: artifactPath }
  }
  if (input.section.required) {
    throw new Error(`Missing required artifact: ${input.section.artifact}`)
  }
  return undefined
}

async function formatArtifactBody(path: string): Promise<string> {
  if (path.endsWith(".json")) {
    return `\`\`\`json\n${JSON.stringify(JSON.parse(await readFile(path, "utf8")), null, 2)}\n\`\`\``
  }
  return `\`\`\`text\n${await readFile(path, "utf8")}\n\`\`\``
}

function displayPath(path: string, repoRoot: string): string {
  const relativePath = relative(repoRoot, resolve(path))
  return relativePath.startsWith("..") ? path : relativePath
}

function findSection(stage: StageContract, label: string, stageId: string): InjectionSection {
  const section = stage.injection.sections.find((candidate) => candidate.label === label)
  if (section === undefined) {
    throw new Error(`Missing injection section '${label}' for stage '${stageId}'`)
  }
  return section
}

File: templates/production-adapters/opencode-http-session/python/prompt_assembly.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

from contracts import InjectionSection, StageContract, WorkflowManifest


WORKFLOW_LOCAL_PREFIXES = ("rules/", "prompts/", "reference/")


@dataclass(frozen=True, slots=True)
class PromptRender:
    stage: str
    text: str
    prompt_path: Path


def render_prompt(
    *,
    stage: str,
    workflow_root: Path,
    run_dir: Path,
    attempt: int,
    repo_root: Path,
    manifest: WorkflowManifest,
    input_artifact: str,
) -> PromptRender:
    stage_contract = manifest.stages[stage]
    prompt_template_path = workflow_root / stage_contract.prompt_template
    sections = [
        render_runner_context(
            stage,
            stage_contract=stage_contract,
            run_dir=run_dir,
            repo_root=repo_root,
        )
    ]
    selected_artifacts: list[dict[str, str]] = []
    for label in stage_contract.injection.render_order:
        section = _find_section(stage_contract, label=label, stage=stage)
        rendered = read_section_body(
            section=section,
            prompt_template_path=prompt_template_path,
            workflow_root=workflow_root,
            run_dir=run_dir,
            repo_root=repo_root,
        )
        if rendered is None:
            continue
        body, resolved_path = rendered
        sections.append(f"# {section.label}\n\n{body}")
        if resolved_path is not None:
            selected_artifacts.append({"label": section.label, "path": display_path(resolved_path, repo_root=repo_root)})
    if stage_contract.continuation_message is not None:
        sections.append("# Continuation Message\n\n" + stage_contract.continuation_message)
    prompt_path = run_dir / "artifacts" / "prompts" / f"{stage}-{attempt}.md"
    prompt_path.parent.mkdir(parents=True, exist_ok=True)
    prompt_text = "\n\n".join(section for section in sections if section.strip()) + "\n"
    prompt_path.write_text(prompt_text, encoding="utf-8")
    selected_path = run_dir / stage_contract.persistence.selected_artifacts_record
    selected_path.parent.mkdir(parents=True, exist_ok=True)
    selected_path.write_text(json.dumps(selected_artifacts, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    return PromptRender(stage=stage, text=prompt_text, prompt_path=prompt_path)


def ensure_stage_outputs(
    stage: str,
    *,
    run_dir: Path,
    repo_root: Path,
    workflow_root: Path,
    manifest: WorkflowManifest,
) -> None:
    missing = [
        path
        for path in required_output_paths(stage, run_dir=run_dir, repo_root=repo_root, workflow_root=workflow_root, manifest=manifest)
        if not path.exists()
    ]
    if missing:
        raise RuntimeError("stage " + stage + " is missing required outputs: " + ", ".join(str(path) for path in missing))


def required_output_paths(
    stage: str,
    *,
    run_dir: Path,
    repo_root: Path,
    workflow_root: Path,
    manifest: WorkflowManifest,
) -> list[Path]:
    stage_contract = manifest.stages[stage]
    return [
        resolve_contract_path(spec, workflow_root=workflow_root, run_dir=run_dir, repo_root=repo_root)
        for spec in stage_contract.required_outputs
    ]


def clear_stage_outputs(
    stage: str,
    *,
    run_dir: Path,
    repo_root: Path,
    workflow_root: Path,
    manifest: WorkflowManifest,
) -> None:
    stage_contract = manifest.stages[stage]
    for spec in stage_contract.clear_on_enter:
        path = resolve_contract_path(spec, workflow_root=workflow_root, run_dir=run_dir, repo_root=repo_root)
        if path.is_dir():
            continue
        if path.exists():
            path.unlink()


def resolve_contract_path(spec: str, *, workflow_root: Path, run_dir: Path, repo_root: Path) -> Path:
    if spec.startswith(WORKFLOW_LOCAL_PREFIXES):
        return workflow_root / spec
    candidate = Path(spec)
    if candidate.is_absolute():
        return candidate
    return run_dir / spec


def render_runner_context(stage: str, *, stage_contract: StageContract, run_dir: Path, repo_root: Path) -> str:
    outputs = "\n".join(f"- {relative}" for relative in stage_contract.required_outputs)
    return (
        "## Runner Context\n\n"
        f"- Stage key: `{stage}`\n"
        f"- Repository root: `{repo_root}`\n"
        f"- Run directory: `{run_dir}`\n"
        "- Required outputs:\n"
        f"{outputs}"
    )


def read_section_body(
    *,
    section: InjectionSection,
    prompt_template_path: Path,
    workflow_root: Path,
    run_dir: Path,
    repo_root: Path,
) -> tuple[str, Path | None] | None:
    if section.prompt_template:
        return prompt_template_path.read_text(encoding="utf-8"), None
    if section.artifact is None:
        raise RuntimeError(f"Section '{section.label}' must define artifact or promptTemplate")
    artifact_path = resolve_contract_path(section.artifact, workflow_root=workflow_root, run_dir=run_dir, repo_root=repo_root)
    if artifact_path.exists():
        return format_artifact_body(artifact_path), artifact_path
    if section.required:
        raise RuntimeError(f"Missing required artifact: {section.artifact}")
    return None


def format_artifact_body(path: Path) -> str:
    if path.suffix == ".json":
        payload = json.loads(path.read_text(encoding="utf-8"))
        body = json.dumps(payload, indent=2, sort_keys=True)
        return f"```json\n{body}\n```"
    return f"```text\n{path.read_text(encoding='utf-8')}\n```"


def display_path(path: Path, *, repo_root: Path) -> str:
    try:
        return str(path.relative_to(repo_root))
    except ValueError:
        return str(path)


def _find_section(stage_contract: StageContract, *, label: str, stage: str) -> InjectionSection:
    for section in stage_contract.injection.sections:
        if section.label == label:
            return section
    raise RuntimeError(f"Missing injection section '{label}' for stage '{stage}'")
